File: backend/services/_archived/identity/src/services/RbacService.ts
import type { UserRole } from '../domain/User.js';

export const ROLE_LEVEL: Record<UserRole, number> = {
  SUPER_ADMIN: 100,
  DEPT_ADMIN: 80,
  VC: 70,
  REGISTRAR: 70,
  DEAN: 60,
  HOD: 50,
  STAFF: 10,
};

export class RbacService {
  /**
   * Returns true if the user's role is in the list of required roles,
   * OR if the user's role level exceeds all required role levels
   * (i.e. SUPER_ADMIN can always access).
   */
  canAccess(userRole: UserRole, requiredRoles: UserRole[]): boolean {
    if (requiredRoles.length === 0) return true;

    // Direct match
    if (requiredRoles.includes(userRole)) return true;

    // Hierarchy: user's level must meet or exceed ALL required roles' levels
    const userLevel = ROLE_LEVEL[userRole];
    const maxRequiredLevel = Math.max(...requiredRoles.map((r) => ROLE_LEVEL[r]));

    return userLevel >= maxRequiredLevel;
  }
}

File: backend/api/src/modules/identity/services/rbac_service.py
"""Role-Based Access Control service.

Role hierarchy (higher level = more access):
  SUPER_ADMIN (100) > DEPT_ADMIN (80) > VC (70) = REGISTRAR (70) > DEAN (60) > HOD (50) > STAFF (10)
"""

ROLE_LEVELS: dict[str, int] = {
    "SUPER_ADMIN": 100,
    "DEPT_ADMIN": 80,
    "VC": 70,
    "REGISTRAR": 70,
    "DEAN": 60,
    "HOD": 50,
    "STAFF": 10,
}


class RbacService:
    def can_access(self, user_role: str, required_roles: list[str]) -> bool:
        """Return True if user_role matches or exceeds all required roles."""
        user_level = ROLE_LEVELS.get(user_role, 0)
        for role in required_roles:
            required_level = ROLE_LEVELS.get(role, 0)
            if user_level < required_level:
                return False
        return True
